src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeoAuthProvider.scala

```diff
@@ -25,45 +24,38 @@ object NotebookClusterActions {
     */
-  def hasProjectPermission(userEmail: WorkbenchEmail, action: ProjectActions.ProjectAction, googleProject: GoogleProject)(implicit executionContext: ExecutionContext): Future[Boolean]
-
-  /**
-    * When listing clusters, Leo will perform a GROUP BY on google projects and call this function once per google project.
-    * If you have an implementation such that users, even in some cases, can see all clusters in a google project, overriding
-    * this function may lead to significant performance improvements.
-    * For any projects where this function call returns Future.successful(false), Leo will then call hasNotebookClusterPermission
-    * for every cluster in that project, passing in action = GetClusterStatus.
-    *
-    * @param userEmail The user in question
-    * @param googleProject A Google project
-    * @return If the given user can see all clusters in this project
-    */
-  def canSeeAllClustersInProject(userEmail: WorkbenchEmail, googleProject: GoogleProject)(implicit executionContext: ExecutionContext): Future[Boolean] = {
-    Future.successful(false)
-  }
+  def hasProjectPermission(userInfo: UserInfo, action: ProjectActions.ProjectAction, googleProject: GoogleProject)(implicit executionContext: ExecutionContext): Future[Boolean]
 
   /**
     * Leo calls this method to verify if the user has permission to perform the given action on a specific notebook cluster.
     * It may call this method passing in a cluster that doesn't exist. Return Future.successful(false) if so.
     *
-    * @param userEmail      The user in question
+    * @param userInfo      The user in question
     * @param action        The cluster-level action (above) the user is requesting
     * @param googleProject The Google project the cluster was created in
     * @param clusterName   The user-provided name of the Dataproc cluster
     * @return If the userEmail has permission on this individual notebook cluster to perform this action
     */
-  def hasNotebookClusterPermission(userEmail: WorkbenchEmail, action: NotebookClusterActions.NotebookClusterAction, googleProject: GoogleProject, clusterName: ClusterName)(implicit executionContext: ExecutionContext): Future[Boolean]
+  def hasNotebookClusterPermission(userInfo: UserInfo, action: NotebookClusterActions.NotebookClusterAction, googleProject: GoogleProject, clusterName: ClusterName)(implicit executionContext: ExecutionContext): Future[Boolean]
+
+  /**
+    * Leo calls this method when it receives a "list clusters" API call, passing in all non-deleted clusters from the database.
+    *
+    * @param userInfo The user in question
+    * @param clusters All non-deleted clusters from the database
+    * @return         Filtered list of clusters to return to the caller
+    */
+  def filterClusters(userInfo: UserInfo, clusters: List[(GoogleProject, ClusterName)])(implicit executionContext: ExecutionContext): Future[List[(GoogleProject, ClusterName)]]
```

Comment:
I got rid of `canSeeAllClustersInProject` and added this method as a way to support listing clusters. Filtering a list of clusters seems like a reasonable thing for a provider to do, and makes it more flexible too.

Reply:
You probably want to call this `filterUserVisibleClusters` or something, and indicate in the comment that you should return a list of clusters that the user can see according to their authz. I don't think it's clear from either the function name or the comment (filtered how?).

Reply:
👍